Add `_change_block` to entities GraphQL `input` type filter

dotansimha opened the pull request:
Related: https://github.com/graphprotocol/graph-node/issues/1838 

### Background

The initial discussion started at https://github.com/graphprotocol/graph-node/issues/2958 about adding a property to every type, so users can filter by it. It brought back https://github.com/graphprotocol/graph-node/issues/1838 where a user should be able to filter by block.

### Overview of the changes

* Added `EntityFilter::ChangeBlockGet` and implemented the SQL required by it.
* Added a new `input BlockChangedFilter` to allow filtering based on `number_gte` of the block change.
* No GraphQL breaking changes.
* Added special handling for filter key `_change_block` and translate that into the SQL query. 

### TODO:
- [x] Figure out the GraphQL schema structure 
- [x] Add field filter to the GraphQL `input` named `_change_block` 
- [x] Change HEAD to be based on @lutter 's refactor PR 
- [x] Introduce new input type for filtering
- [x] Modify resolvers/SQL/prefetch execution

![image](https://user-images.githubusercontent.com/3680083/145010226-db114311-1f41-4250-9608-e60403f745e9.png)

![image](https://user-images.githubusercontent.com/3680083/145010275-abd8c803-e85c-4850-b5ec-580755eb8bc2.png)

yanivtal commented:
Awesome! I had pointed this old issue out to Adam for what we originally had in mind for what we called the [Confirmations API](https://github.com/graphprotocol/graph-node/issues/295). Have you had a chance to look at it and how would you compare the API proposed here vs the one off field in this design?

dotansimha commented:
> Awesome! Great to see this PR. I had pointed this old issue out to Adam for what we originally had in mind for what we called the [Confirmations API](https://github.com/graphprotocol/graph-node/issues/295). Have you had a chance to look at it and how would you compare the API proposed here vs the one off field in this design?

Thanks @yanivtal , this is super interesting. 

I think in terms of comparison of the two features, I can only say that Confirmations API seems more comprehensive, and exposes information about the actual changes, and not just the data itself. 
In this PR, I added only a filter for `_change_block` that translated into a simple `>=` operation in the database and doesn't expose any additional information (so it acts just like a basic filter). 

I went through the issue you mentioned, and it seems like that specific API that I added in this PR is not part of the examples? Or did I miss it? the actual change here allows you to fetch `tokens` that are changed after a specific block (at the level of `tokens` field), and not the information about the change. 
I guess Confirmations API can be an alternative for exposing metadata information as suggested in https://github.com/graphprotocol/graph-node/issues/2958 ? 

With some addions to Confirmations GraphQL API, I think we can introduce a similar filter on a higher level, something like: 

```graphql
schema {
  query: Query
}

type Query {
  _block(filter: BlockFilter!): EntitiesRootQuery!
  tokens(...): [Tokens]
}

input BlockFilter {
   gte: Int
}

type EntitiesRootQuery {
  tokens(...): [Tokens]
}
```

And then users can inspect the data from the point of view of a block (or, block filter), and fetch the data it needs - so the `tokens` filter remains "clear" and deals only with the actual entities, and not with the block information? 

```graphql
query {
  _block(gte: 10) {
     tokens(..) { 
        id
     }
  }
}
```

what do you think? (just a thought, I guess this might be more relevant to the discussion on GraphQL API v2?)

dotansimha commented on graphql/src/schema/api.rs:
```diff
@@ -25,6 +25,7 @@ pub enum APISchemaError {
 }
 
 const BLOCK_HEIGHT: &str = "Block_height";
+const CHANGE_BLOCK_FILTER_NAME: &str = "BlockChangedFilter";
```
@lutter I saw that we don't have a clear naming convention standard for the built GraphQL schema, I used PascalCase because that's what we are using for inputs.

lutter replied:
You are right - it seems that most other type names use PascalCase; I might have screwed that up in the past when I added `Block_height` instead of `BlockHeight`; I think using that name is good and continues the more prevalent pattern.

azf20 commented:
> how would you compare the API proposed here vs the one off field in this design?

This PR is focused on filtering entities to support two specific use cases - fetching new entities as they are indexed, and debugging / inspecting when entities changed. The Confirmations API is much more feature rich, with a much bigger impact on the overall generated schema, surfacing field level changes, as well as entity level changes. That would require significant changes in SQL generation and perhaps database changes, while this was prioritised as a useful, but more incremental, feature

yanivtal commented:
Okay I'm fine with whatever you guys want to do here. Just want to make sure that as we consider API design, we're cognizant of the effect of adding many one off fields vs thinking through the structure of the underlying data model more holistically. If a one off field is more appropriate here I support.

lutter commented on graphql/src/schema/api.rs:
```diff
@@ -240,6 +242,27 @@ fn add_block_height_type(schema: &mut Document) {
 
+/// Adds a global `BlockChangedFilter` type to the schema. The `block` argument
+/// accepts values of this type
+fn add_changed_block_filter_input_type(schema: &mut Document) {
```
This is not something you need to change, but directionally, I don't like how we create all these fixed types in code, just because it makes understanding what gets added much harder. With `meta.graphql` I started defining these types in GraphQL text, which I think is much easier to read than constructing an AST in code.

If we ever make bigger changes to the API schema in the future, I think it would be great to rename `meta.graphql` and add to that whatever can be added with fixed GraphQL text.

dotansimha replied:
Yeah, I agree that the SDL in `meta.graphql` is more readable. 
I will move the definition of `BlockChangedFilter` and `Block_height`.

lutter commented on graphql/src/schema/api.rs:
```diff
@@ -240,6 +242,27 @@ fn add_block_height_type(schema: &mut Document) {
+            position: Pos::default(),
+            description: None,
+            name: "number_gte".to_owned(),
+            value_type: Type::NamedType("Int".to_owned()),
```
That should be `number_gte: Int!`, not just `number_gte: Int`, right?

dotansimha replied:
Yeah, that's right :) will fix and move to `meta.graphql`.

lutter commented on graphql/src/store/query.rs:
```diff
@@ -141,6 +141,24 @@ fn build_fulltext_filter_from_object(
 }
 
+fn parse_change_block_filter(
+    value: &graph::data::value::Value,
```
It's minor, but since we have so many `Value` structs floating around worth to be consistent: most of the code that deals with `graph::data::value::Value` actually does a `use graph::prelude::r` and then calls it `r::Value` (the `r` is because we call values from `graphql_parser` `q::Value` and `r` is the next letter ;) )

lutter commented on graphql/src/store/query.rs:
```diff
@@ -141,6 +141,24 @@ fn build_fulltext_filter_from_object(
+                .ok_or_else(|| QueryExecutionError::InvalidFilterError)?;
+
+            match block_value {
+                graph::data::value::Value::Int(n) => Ok(*n as i32),
```
There's a helper for this conversion: use `i32::try_from_value(block_value).map_err(|_| QueryExecutionError::InvalidFilterError)` instead of the `match` statement to do a conversion that checks bounds.

dotansimha replied:
agree, will refactor :)

lutter commented on store/postgres/src/relational_queries.rs:
```diff
@@ -1075,6 +1079,21 @@ impl<'a> QueryFilter<'a> {
+        // constructs a filter for lower(c.block_range) >= $block_number_gte
+        // to allow consumers to filter by a specific change block
+        out.push_sql("lower(c.");
+        out.push_identifier("block_range")?;
```
You don't necessarily have to bother with `push_identifier` here since we know that `block_range` is SQL safe; but it would be nice to use the existing constant for that name and make it `out.push_sql(BLOCK_RANGE_COLUMN)`

dotansimha replied:
oops, missed that constant, fixed!

lutter reviewed (changes requested):
Looks really good! Mostly just very small style comments, the only thing that I think needs to be addressed is the conversion to `i32`, though feel free to address other comments, too.  I think this would be good to merge then as soon as we merge #3005

dotansimha commented:
@lutter addressed all PR comments :) Thanks for the code review! 

Any idea why CI fails? the error is:

```
   Compiling wasmtime-obj v0.27.0
     Running `rustc --crate-name wasmtime_obj --edition=2018 /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmtime-obj-0.27.0/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no -C debuginfo=2 -C metadata=23aca09821a9684b -C extra-filename=-23aca09821a9684b --out-dir /home/runner/work/graph-node/graph-node/target/debug/deps -L dependency=/home/runner/work/graph-node/graph-node/target/debug/deps --extern anyhow=/home/runner/work/graph-node/graph-node/target/debug/deps/libanyhow-7ce8467d8e955743.rmeta --extern more_asserts=/home/runner/work/graph-node/graph-node/target/debug/deps/libmore_asserts-bbe2ebcfe7cb672d.rmeta --extern object=/home/runner/work/graph-node/graph-node/target/debug/deps/libobject-233df90730997cd9.rmeta --extern target_lexicon=/home/runner/work/graph-node/graph-node/target/debug/deps/libtarget_lexicon-02e0409bd62baaaf.rmeta --extern wasmtime_debug=/home/runner/work/graph-node/graph-node/target/debug/deps/libwasmtime_debug-ffc768e88aabfe49.rmeta --extern wasmtime_environ=/home/runner/work/graph-node/graph-node/target/debug/deps/libwasmtime_environ-ce368bdfeb8a7ec1.rmeta --cap-lints allow -C link-arg=-fuse-ld=lld -D warnings`
error: field is never read: `logger`
  --> graph/src/firehose/endpoints.rs:21:5
   |
21 |     logger: Logger,
   |     ^^^^^^^^^^^^^^
   |
   = note: `-D dead-code` implied by `-D warnings`
```

but it seems unrelated to the changes I did? 🤔

lutter commented:
Yes, I think that failure was because of the new `rustc` version.

dotansimha commented:
> 

Fixed, and CI passes now, thanks @lutter  :) I think we are good to go with this PR?
